upcoming: [DI-21200] - Migration from ChartJS to Recharts in CloudPulse

nikhagra-akamai opened the pull request:
## Description 📝
Migrated ChartJS to Recharts in CloudPulse.

Note: This PR is raised so that code changes can be reviewed & parallelly we are working on automation test cases which will eventually be added in this PR.

## Changes  🔄
List any change relevant to the reviewer.
1. Replaced LineGraph with AreaChart in CloudPulseLineGraph component.
2. Updated x-axis ticks & chart variant in CloudPulseWidget
3. Modified structure the of data passed to AreaChart component in CloudPulseWidgetUtils
4. Removed CloudPulseColorPalette
5. Modified legendColor property in metrics display to accept any color
6. Added a connectNulls property in AreaChart component
7. Added DataSet type in cloudpulse types

## Target release date 🗓️
18th November

## Preview 📷
**Include a screenshot or screen recording of the change.**

:lock: Use the [Mask Sensitive Data](https://cloud.linode.com/profile/settings) setting for security.

:bulb: Use `<video src="" />` tag when including recordings in table.

No Data Display
![Screenshot 2024-11-04 at 1 41 57 PM](https://github.com/user-attachments/assets/a18d0dd6-025d-41bd-b386-2f1785a6fcca)
![Screenshot 2024-11-04 at 1 42 10 PM](https://github.com/user-attachments/assets/63794c06-9aa7-4539-a356-035806e9b688)

X-Axis with timestamp ticks for less than 24hrs
![Screenshot 2024-11-04 at 1 42 22 PM](https://github.com/user-attachments/assets/7fa7b0aa-a9e4-4aff-92cc-5c7dd8d5162a)
![Screenshot 2024-11-04 at 1 42 30 PM](https://github.com/user-attachments/assets/6731f28a-2a6b-4c7d-b7b1-944c45dfb0f9)

X-Axis with date & timestamp ticks for more than 24hrs
![Screenshot 2024-11-04 at 1 42 57 PM](https://github.com/user-attachments/assets/4e03c854-9ff0-4599-969c-551516ca9761)
![Screenshot 2024-11-04 at 1 43 16 PM](https://github.com/user-attachments/assets/5515a9bd-2b2e-401b-8dc1-0576d4a6aa45)

Legends rows 
https://github.com/user-attachments/assets/bc8886ba-db7a-44ba-aa09-fea7d715bd22

Error State
![Screenshot 2024-11-04 at 2 16 32 PM](https://github.com/user-attachments/assets/c006e94c-055b-4537-b55b-ca5dfe5fd70a)
![Screenshot 2024-11-04 at 2 16 40 PM](https://github.com/user-attachments/assets/122830ba-9917-4e45-aad9-b0e965eb76fd)


## How to test 🧪

1. Switch to mock user
2. Select mandatory filters to generate the graphs
3. Change time duration filter to see difference in x-axis tick
4. Modify data in serverHandlers.ts for `monitor/services/:serviceType/metrics` endpoint to playaround with graph & legends.
5. You can see a scroll bar in legend rows table if more than 2 legends are present



## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [x] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
- [ ] 🧪 Providing/Improving test coverage
- [x] 🔐 Removing all sensitive information from the code and PR description
- [x] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [x] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

abailly-akamai commented:
@nikhagra-akamai please address the e2e failures before we get to the review - thx!

hkhalil-akamai commented on packages/manager/src/features/CloudPulse/Widget/components/CloudPulseLineGraph.tsx:
```diff
@@ -42,25 +33,14 @@ export const CloudPulseLineGraph = React.memo((props: CloudPulseLineGraph) => {
-          legendRows={legendRows}
+          connectNulls
+          fillOpacity={0.5}
+          legendHeight={theme.spacing(18.75)}
```
IMO, the overflow in the legend is worse from a usability perspective -- I would prefer to be able to see all rows without a scroll bar.

![Screenshot 2024-11-04 at 6 06 02 PM](https://github.com/user-attachments/assets/64f69095-1b0e-4295-be3f-3f49b5be8f42)

nikhagra-akamai replied:
This is also already been discussed. I understand your concern, but think for dynamically growing legend rows where number can cross 5. Bigger legend row table won't be look so good for UI and if two side by side graphs have different number of legend row count, then that mismatch in graph & legend height will make it look more worse. As in our case legends rows are dynamic & they can grow to 20.

hkhalil-akamai replied:
Understood, thanks for the context.

nikhagra-akamai commented on packages/manager/src/components/LineGraph/MetricsDisplay.tsx:
```diff
@@ -47,7 +38,7 @@ export interface MetricsDisplayRow {
   format: (n: number) => string;
   handleLegendClick?: () => void;
-  legendColor: LegendColor;
+  legendColor: string;
```
yes it is already been discussed with @jaalah-akamai  that instead of restricting to the 7 predefined colors, it should accept any color, so that legend rows greater than 7 won't collide with same color. And I don't think color has anything related to typesafety.

nikhagra-akamai commented on packages/api-v4/src/cloudpulse/types.ts:
```diff
@@ -132,3 +132,8 @@ export interface ServiceTypes {
+export interface DataSet {
+  [label: string]: number;
+  timestamp: number;
+}
```
updated, thanks

nikhagra-akamai commented on packages/manager/src/components/AreaChart/AreaChart.tsx:
```diff
@@ -64,6 +66,11 @@ interface AreaChartProps {
+  /**
+   * connect nulls value between two data points
+   */
+  connectNulls? :boolean;
```
updated, it was there in local changes but forgot to push

nikhagra-akamai commented:
> @nikhagra-akamai please address the e2e failures before we get to the review - thx!

@abailly-akamai This PR is raised so that code changes can be reviewed & parallelly we are working on automation test cases which will eventually be added in this PR. Let me know if automation is necessary for the review, I'll make it draft & open once automations are done

abailly-akamai commented:
@nikhagra-akamai I am unsure what you meant, but this just can't be merged with failing tests no matter what cause everyone would inherit these failures

abailly-akamai commented:
@nikhagra-akamai fixing the tests may affect the current code and we are trying to minimize the back and forth.

kmuddapo commented:
> @nikhagra-akamai fixing the tests may affect the current code and we are trying to minimize the back and forth.

@nikhagra-akamai  is on OOO for this week.  May be @abailly-akamai / @jaalah  In that case could we please move this PR to draft or even close it? That way once team is ready with E2E automations which is currently in progress, we will open up a consolidated PR.

nikhagra-akamai commented on packages/manager/src/components/AreaChart/AreaChart.tsx:
```diff
@@ -93,6 +106,11 @@ interface AreaChartProps {
    */
   margin?: { bottom: number; left: number; right: number; top: number };
 
+  /**
+   * control the visibility of dots for each data points
+   */
+  showDot?: boolean;
+
```
It is to control the permanent visibility of dots instead of only hover

![Screenshot 2024-11-18 at 12 00 42 PM](https://github.com/user-attachments/assets/664cefb4-935e-4f5f-b8a4-998f279628c0)

nikhagra-akamai commented on packages/manager/src/components/AreaChart/AreaChart.tsx:
```diff
@@ -124,24 +142,33 @@ interface AreaChartProps {
     ariaLabel,
+    connectNulls,
     data,
+    dotRadius = 3,
```
default dotRadius as per the recharts documentation

nikhagra-akamai commented on packages/manager/src/components/AreaChart/AreaChart.tsx:
```diff
@@ -229,9 +256,14 @@ export const AreaChart = (props: AreaChartProps) => {
             dataKey="timestamp"
             domain={['dataMin', 'dataMax']}
-            interval="preserveEnd"
+            interval={xAxisTickCount ? 0 : 'preserveEnd'}
```
@jaalah-akamai this condition is added because preserveEnd will work good for x-axis ticks if we don't want to control it otherwise with 0 we can control it

nikhagra-akamai commented:
@jaalah-akamai / @abailly-akamai / @hkhalil-akamai automation tests are updated with all other changes for recharts migrations. PR is now ready to review & approvals

abailly-akamai reviewed (commented):
This PR can use some cleanup, pointing at strengthening the types a little

additionally, am seeing quite a few differences with the previous implementation in the UI. (horizontal legend is a big improvement 👍 )

![Screenshot 2024-11-18 at 09 37 51](https://github.com/user-attachments/assets/12b5a456-3d46-4ea0-9b39-2f25a73ab87b)

- we now have rounded charts, which is fine but want to confirm UX is on board
- we now display the time in the legend, which makes it unreadable on mobile
![Screenshot 2024-11-18 at 10 09 18](https://github.com/user-attachments/assets/0e25cd8a-5cca-4aff-b8ad-1c48a7d7de30)

Lastly, do we need this feature?
what is the use case for minimizing the chart here?
![Screenshot 2024-11-18 at 10 12 45](https://github.com/user-attachments/assets/5fbf2f8a-a1a5-45a9-bc0b-35e22cba9746)

nikhagra-akamai commented on packages/manager/src/components/AreaChart/utils.ts:
```diff
@@ -36,6 +36,36 @@ export const humanizeLargeData = (value: number) => {
 };
 
+export const generate12HourTicks = (
+  data: any[],
```
this type is tightly coupled with type of data prop in the AreaChart component , and it is previously discussed here may be taken care later. https://github.com/linode/manager/pull/11145#issuecomment-2441642456

nikhagra-akamai replied:
@jaalah-akamai  could you please confirm what needs to be done here as it is similar to what is there in AreaChart's data prop.

nikhagra-akamai replied:
`export interface DataSet {
  [label: string]: number;
  timestamp: number;
}`

@abailly-akamai  @hkhalil-akamai  @jaalah-akamai , if there is issue with `any` then we can use this `DataSet` type which is same as the type of data expected in AreaChart component.

abailly-akamai replied:
@nikhagra-akamai consider using `any` as **always** being an issue, and blocking your PRs.

If `DataSet` is applicable and generally accepted as data passed to the Charts, then use, please use that

nikhagra-akamai replied:
Sure will update it with DataSet

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Widget/CloudPulseWidget.tsx:
```diff
@@ -238,11 +238,12 @@ export const CloudPulseWidget = (props: CloudPulseWidgetProperties) => {
+  let legendRows: MetricsDisplayRow[] = [];
+  let currentUnit = unit;
+  let areas: AreaProps[] = [];
+  const variant: ChartVariant = widget.chart_type === 'line' ? 'line' : 'area';
```
this chart_type property is coming from API response as part of widget configuration and this name is as per API spec.

abailly-akamai replied:
API specs: `chart_type enum ('line','area'),`

It's a simple, logical change, straight from your own specs. I don't see a reason not to [add it](https://github.com/linode/manager/blob/develop/packages/api-v4/src/cloudpulse/types.ts#L31) to reflect a true type.

bnussman-akamai replied:
I agree. 

If the API spec clearly defines `chart_type` as an enum, it would be good if we update our `api-v4` package to align:

```typescript
chart_type: 'line' | 'area';
```

nikhagra-akamai replied:
okay, I misunderstood it, I thought @abailly-akamai  was pointing at the property name. Will update thanks for suggestion

kmuddapo commented:
> additionally, am seeing quite a few differences with the previous implementation in the UI. (horizontal legend is a big improvement 👍 )
@abailly-akamai  , There is still an open issue with horizontal legend, you could notice when you bigger legend  labels, which Jaalah is looking into on via [M3-8872](https://jira.linode.com/browse/M3-8872),

kmuddapo commented:
> * we now have rounded charts, which is fine but want to confirm UX is on board
@abailly-akamai   I hope rounded here you are referring is on dots. If so, yes!! this is approved by the UX.

nikhagra-akamai commented on packages/manager/src/components/LineGraph/MetricDisplay.styles.ts:
this is part of LineGraph package for so long, do we really need to modify the file name now?

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Utils/CloudPulseWidgetUtils.ts:
```diff
@@ -378,7 +403,7 @@ export const fillMissingTimeStampsAcrossDimensions = (
     return sortedTimestamps.map((key) => {
       // If the current array has the key, use its value; otherwise, set it to null, so that the gap is properly visible
-      return [key, map.get(key) ?? null] as [number, number | null];
+      return [key, map.get(key) ?? null] as [number, null | number];
```
this is a temporary fix as part of chartjs, it is removed now

nikhagra-akamai commented:
> In the latest revision, only two rows fit into the viewport at a time. Do we have any estimates on how many rows were expecting on average? If there are 5+ rows, why are we restricting the window to only show 2 instead of 4 or more?

there is no surety how many rows will be there on average it may vary from 1 to 20. And restricting it to 2 is as per mockups we received.

nikhagra-akamai commented:
> we now display the time in the legend, which makes it unreadable on mobile
 
 thank you for highlighting

https://github.com/user-attachments/assets/a86e87c7-237d-4ca8-9f73-35a71f076a7d

nikhagra-akamai commented on packages/manager/src/components/AreaChart/AreaChart.tsx:
```diff
@@ -229,9 +256,14 @@ export const AreaChart = (props: AreaChartProps) => {
             dataKey="timestamp"
             domain={['dataMin', 'dataMax']}
-            interval="preserveEnd"
+            interval={xAxisTickCount ? 0 : 'preserveEnd'}
```
the reason I put that condition is because, the AreaChart component is generating chart with timeseries data & showing 0 x-axis ticks will not provide any insights about the graph specially if data connected with time. What do you think?

abailly-akamai replied:
I think this works with original code or suggestion. 

if (`xAxisTickCount ` is `0` or `undefined`) show `0`

Is that the intent?

nikhagra-akamai replied:
Here my intention is, if someone is passing undefined or 0 to `xAxisTickCount`, then with 'preserveEnd' on interval prop, recharts will generate x axis ticks automatically or else we will generate with `generate12HourTicks` function by passing 0 to interval prop.

And I think, empty x axis will not give any insight about the graph,  that's why I have considered it as a false in the condition instead of separate undefined check.

May be we can discuss on this if required.

abailly-akamai replied:
This may need to be documented in code. 

I was wrong in my previous comment, maybe this just needs to be more explicit (and add a comment)

ie: do you want to handle it the same way if `xAxisTickCount` is either `0` or `undefined`?

nikhagra-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/dbaas-widgets-verification.spec.ts:
```diff
@@ -101,28 +102,49 @@ const metricsAPIResponsePayload = cloudPulseMetricsResponseFactory.build({
+      },
+    ],
+    serviceType: serviceType,
+    status: 'success',
```
updated based on react query
![Screenshot 2024-11-19 at 4 01 07 PM](https://github.com/user-attachments/assets/67e70fd0-7dec-43ab-9d45-5d6c73efb7a1)

nikhagra-akamai commented on packages/manager/cypress/e2e/core/cloudpulse/dbaas-widgets-verification.spec.ts:
```diff
@@ -101,28 +102,49 @@ const metricsAPIResponsePayload = cloudPulseMetricsResponseFactory.build({
-  const roundedMax = Math.round(max * 100) / 100;
+  // Generate graph data using the provided parameters
+  const graphData = generateGraphData({
+    flags: { enabled: true } as Partial<Flags>,
```
updated thanks

abailly-akamai commented on packages/api-v4/src/cloudpulse/types.ts:
```diff
@@ -39,7 +39,7 @@ export interface Widgets {
   color: string;
   size: number;
-  chart_type: string;
+  chart_type: 'line' | 'area';
```
Thank you

abailly-akamai reviewed (approved):
Thanks for strengthening the types )it will go a long way!) and addressing the mobile defects. This migration is important, appreciate the tedious work pushing this forward, charts are never easy 👍 

Approving pending some extra cleanup on the `data` types.

nikhagra-akamai commented:
Thanks @abailly-akamai for the approval

hkhalil-akamai reviewed (approved):
Thanks for responding to feedback!

nikhagra-akamai commented:
thanks @abailly-akamai @hkhalil-akamai for the approvals

couple of updates from the last few comments

1. Renamed the filename `MetricDisplay.styles.ts` to `MetricsDisplay.styles.ts`
2. Added a `DataSet` interface in AreaChart component & used the same of type safety in `generate12HoursTick` function
3. Added more detailed documentation for `xAxisTickCountProperty` 


if all good, lets merge
